Isabelle Backend for Vehicle

samysweb opened the pull request:
I thought I'd open a pull request so that people can look at the current state more easily if interested.
This is still work in progress but has come along pretty far now.

**Remaining ToDos:**
- [x] Compilation for records (Isabelle has records, so this should be straightforward)
- [x] Refactor proof for Wind Controller (there were some changes after completing the initial version)

samysweb committed:
(Manual) Isabelle prototype

samysweb committed:
first steps for automated isabelle generation, still incomplete

samysweb committed:
some progress

samysweb committed:
First attempt in getting Abs/Rep to work in some of the right places

samysweb committed:
Wind seems to compile, have to check utility

samysweb committed:
Current state of Isabelle generation (and what is still missing)

samysweb committed:
Cleaned up Isabelle

samysweb committed:
Refactoring

samysweb committed:
bug fixes

samysweb committed:
Now supporting zero dim tensors

samysweb committed:
Added automatic generation of type rewrite rule

samysweb committed:
some compatibility improvements

samysweb committed:
Update for Vehicle to Isabelle compilation: Everything but Records seems to compile on the Isabelle side

samysweb committed:
Minor change in Tasty Golden Tests:

Allow for copying of directories (necessary for accessing the vehicle-isabelle package during IsabelleVerify golden tests)

samysweb committed:
Comprehensive golden tests for Isabelle compilation

samysweb committed:
Records and fix for pointwise min/max

samysweb committed:
Disabled Isabelle integration test

samysweb committed:
Fixed simplification groups for tensor operations (Vehicle Isabelle Lib)

samysweb committed:
Fixed safety proof for updated version of Wind controller

samysweb committed:
Some first golden tests (may need to fix after rebase)

samysweb committed:
Updated tests after rebase

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

samysweb committed:
Isabelle documentation

samysweb committed:
Merge branch 'samysweb/vehicle-isabelle' of github.com:samysweb/vehicle into samysweb/vehicle-isabelle

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

samysweb commented:
Rebased to the current version and ready for review @MatthewDaggitt

I managed to get the tensor example to work. All golden tests are compilable in Isabelle (I've also added disabled golden tests, which check whether Isabelle compiles the output of Vehicle).

MatthewDaggitt commented on docs/exporting.rst:
```diff
@@ -138,3 +138,28 @@ directly interface with mathcomp's structure hierarchy. This can lead to issues
+No integration with verification cache
+######################################
+There is currently no integration with Vehicle's verification cache,
+meaning that it is up to the user to garuntee that the compiled specification
```
guarantee

samysweb replied:
fixed

MatthewDaggitt commented on vehicle-isabelle/.gitignore:
```diff
@@ -0,0 +1 @@
+*~
```
Rather than adding lots of extra `.gitignore` files scattered throughout, can we add these to the top-level .gitignore file?

samysweb replied:
fixed

MatthewDaggitt commented on vehicle/tests/golden/compile/andGate/test.json:
```diff
@@ -23,6 +23,32 @@
+  {
+    "name": "IsabelleVerify",
+    "run": "vehicle compile itp -s spec.vcl -t Isabelle -o isabelle/Isabelle.thy && isabelle build -d vehicle-isabelle/ -d isabelle/ IsabelleTest > /dev/null 2>&1 && echo $? || echo 1",
+    "enabled": false,
```
These should be enabled, but to stop then running when issabelle is not available, you can add the `external` field (see the Rocq example above).

samysweb replied:
fixed

MatthewDaggitt commented on vehicle/tests/golden/compile/andGate/isabelle/ROOT:
```diff
@@ -0,0 +1,7 @@
+session IsabelleTest = HOL +
```
Are these ROOT files needed to be generated manually for each test?

samysweb replied:
Alternatively, I could create the ROOT file once somewhere and always copy it to the test directory... Would you prefer that?

samysweb replied:
To put this somewhere else, I would need some place to put an `isabelle` directory with a `ROOT` file which I can then provide in the golden test's `"needs"` section (this would copy `isabelle/ROOT` into the temporary test directory) -- would you have a preferred place for this?

MatthewDaggitt replied:
No, that's fine. If they are needed then you can leave them here.

MatthewDaggitt commented on vehicle/src/Vehicle/Backend/Isabelle/Compile.hs:
```diff
@@ -0,0 +1,897 @@
+
+currentPhase :: Doc ()
+currentPhase = "compilation to Isabelle"
+compileProgToIsabelle :: (MonadCompile m) => Prog DecidabilityBuiltin -> IsabelleOptions -> m (Doc a)
```
Is there a reason the formatter isn't automatically putting an empty line between `currentPhase` and `compileProgToIsabelle`?

samysweb replied:
I don't know, but I added a new line now (fixed)

MatthewDaggitt commented on vehicle/src/Vehicle/Backend/Isabelle/Compile.hs:
```diff
@@ -0,0 +1,897 @@
+  logCompilerSection2 MinDetail currentPhase $ do
+    logDebug MaxDetail $ prettyExternal (Main ds)
+    -- Combine the printed documents
+    -- let combined = fold gatheredStatements
```
Can remove this unneeded code?

samysweb replied:
fixed (for all comments)

MatthewDaggitt commented on vehicle/src/Vehicle/Backend/Isabelle/Compile.hs:
```diff
@@ -0,0 +1,897 @@
+--   and folding all dependency annotations.
+collectCodeDependencies :: Code -> Set Dependency
+collectCodeDependencies doc =
+  let stream = layoutPretty defaultLayoutOptions doc
```
It's usually cleaner to write lets like this using do notation, e.g.
```
collectCodeDependencies doc = do
  let stream = layoutPretty defaultLayoutOptions doc
  fold (reAnnotateS fst stream)
```

samysweb replied:
fixed

MatthewDaggitt commented on vehicle/src/Vehicle/Backend/Isabelle/Compile.hs:
```diff
@@ -0,0 +1,897 @@
+collectLocaleDependencies = Set.unions . fmap deps
+  where
+    deps :: LocaleDef -> Set Dependency
+    deps (NetworkDefStatement name ty) =
```
Usually cleaner to use `\case`, i.e.
```
deps = \case
    NetworkDefStatement name ty ->  collectCodeDependencies name `Set.union` collectCodeDependencies ty
    PropertyDefStatement stmt -> collectCodeDependencies stmt
    TensorTypeDefStmt _ shape body -> collectCodeDependencies shape `Set.union` collectCodeDependencies body
    IndexTypeDefStmt _ maxI body -> collectCodeDependencies maxI `Set.union` collectCodeDependencies body
```
Then you can rename the definition without having to touch so many lines of code.

samysweb replied:
fixed (for all occurrences)

MatthewDaggitt commented on vehicle/src/Vehicle/Backend/Isabelle/Compile.hs:
```diff
@@ -0,0 +1,897 @@
+  = NetworkDefStatement Code Code
+  | PropertyDefStatement Code
+  | TensorTypeDefStmt Identifier Code Code
+  | IndexTypeDefStmt Identifier Code Code
```
Can we have empty lines between declarations please?

samysweb replied:
fixed (hopefully everywhere)

MatthewDaggitt commented on vehicle/src/Vehicle/Backend/Isabelle/Compile.hs:
```diff
@@ -0,0 +1,897 @@
+
+gatherLocaleStatements :: (MonadIsabelleCompile m) => IsabelleOptions -> [LocaleDef] -> Decl DecidabilityBuiltin -> m [LocaleDef]
+gatherLocaleStatements _opts localeNets = \case
+  -- DefFunction _ n _ (Universe _ _) e -> do
```
Delete unneeded code?

samysweb replied:
fixed (for all comments)

MatthewDaggitt commented on vehicle/src/Vehicle/Backend/Isabelle/Compile.hs:
```diff
@@ -0,0 +1,897 @@
+      else compileFunDef localeAssms n t binders body
+        --bindersT bindersV cbody
+  DefRecord _ n _ _t fs -> do
+    -- t' <- compileExpr False localeAssms t
```
Unneeded code?

samysweb replied:
fixed (for all comments)

MatthewDaggitt commented on vehicle/src/Vehicle/Backend/Isabelle/Compile.hs:
```diff
@@ -0,0 +1,897 @@
+  result <- case expr of
+    Hole {} -> resolutionError currentPhase "Hole"
+    Meta {} -> resolutionError currentPhase "Meta"
+    Universe _ l -> developerError $ "compilation of universes to Isabelle unsupported: " <> pretty l
```
Should this call `compileType` instead?

samysweb replied:
I've unified the error message to compileType now...

MatthewDaggitt commented on vehicle/src/Vehicle/Backend/Isabelle/Compile.hs:
```diff
@@ -0,0 +1,897 @@
+    Lam _ binder body -> compileLam localeAssms binder body
+    Builtin _p b -> compileBuiltin isOutType localeAssms b []
+    App fun args -> compileApp isOutType localeAssms fun args
+    Record _p _i fs -> do
```
This is quite complicated. Can we split it out into a `compileRecord` function?

samysweb replied:
This is still on my TODO list

samysweb replied:
done

MatthewDaggitt commented on vehicle/src/Vehicle/Backend/Isabelle/Compile.hs:
```diff
@@ -0,0 +1,897 @@
+    else return $ localeResults ++ funBinders
+
+compileLocaleBindersT :: [LocaleDef] -> [Code]
+compileLocaleBindersT =  (mapMaybe compileLocaleBindersTMapper)
```
Brackets not needed.

samysweb replied:
fixed

MatthewDaggitt commented on vehicle/src/Vehicle/Backend/Isabelle/Compile.hs:
```diff
@@ -0,0 +1,897 @@
+compileLocaleBindersT :: [LocaleDef] -> [Code]
+compileLocaleBindersT =  (mapMaybe compileLocaleBindersTMapper)
+  where
+    compileLocaleBindersTMapper (NetworkDefStatement _ t) = Just (parens t)
```
Can we use `\case` here and below.

samysweb replied:
fixed

MatthewDaggitt commented on vehicle/src/Vehicle/Backend/Isabelle/Compile.hs:
```diff
@@ -0,0 +1,897 @@
+  return $ pretty field <+> "=" <+> (parens fieldValue')
+
+compileTensorTypeDef :: Identifier -> Code -> Code -> Code
+compileTensorTypeDef n shape e = ((vsep :: [Code] -> Code) [
```
Shouldn't need the type annotations on `vsep`?

samysweb replied:
I tried removing the type annotation, but this produces a type error -- I'm not sure why

MatthewDaggitt commented on vehicle/src/Vehicle/Backend/Isabelle/Compile.hs:
```diff
@@ -0,0 +1,897 @@
+  where
+    toTensor :: TensorShape -> [Code] -> Code
+    toTensor _ _ = developerError $
+      "Tensor literal compilation not implemented yet" -- TODO: re-enable when we support tensor literals
```
This feels like a big flaw as tensor literals are used quite a lot. Is there some way we can fake them?

samysweb replied:
Good thing you are reminding me of this: I meant to come back to this when I get the error when running the golden tests, but this never happened.
`compileTensorLiteral` is called for `NatTensorLiteral`/`BoolTensorLiteral`/`RatTensorLiteral` (`StandardBuiltinConstructor`)
But there is also `ConstTensor` (`StandardBuiltinFunction`), which seems to correctly compile tensors (line 653).
It would maybe/probably make sense to extract this to the compileTensorLiteral functions, but can you give me an example of what I need to write in a vehicle spec file to end up in the `NatTensorLiteral`/`BoolTensorLiteral`/`RatTensorLiteral`  case? All the cases from the golden tests seem to end up in the `ConstTensor` case

MatthewDaggitt replied:
```haskell
x : Tensor Rat [2]
x = [1.0, 2.0]
```
should end up with a `RatTensorLiteral` for the value of `x

samysweb replied:
So the original version gave me a type error (I'll attach it below).
I now tried out:
```
x : Tensor Real [2]
x = [1.0, 2.0]

y : Tensor Bool [2]
y = [True, False]

numberOfDogs = 6

type Dog = Index numberOfDogs

z : List Dog
z = [1, 2]
```
This yields the following (which works):
```Isabelle
definition x  :: "  R FlexTensor "
  where " x   = ( (combine_subtensors [(flextensor_from_vec [] [ (((1 :: R) )) ]),(flextensor_from_vec [] [ (((2 :: R) )) ])]) ) "

definition y  :: "  bool FlexTensor "
  where " y   = ( (combine_subtensors [(flextensor_from_vec [] [ true ]),(flextensor_from_vec [] [ false ])]) ) "

definition numberOfDogs  :: "  nat "
  where " numberOfDogs   = ( (6 :: nat) ) "

typedef Dog  =  "{ i :: nat. i < (numberOfDogs) }"
unfolding numberOfDogs_def
  by (simp, rule_tac x = "0" in exI, linarith)

(* Type Coercions *)
declare [[coercion Rep_Dog]]
definition to_Dog :: "FlexIndex \<Rightarrow> Dog"
  where[simp]: "to_Dog a = (
    let i = Rep_FlexIndex a
    in (if i < numberOfDogs then Abs_Dog i else undefined))"
declare [[coercion to_Dog]]

definition z  :: "  Dog list "
  where " z   = ( (Abs_FlexIndex 1) # ((Abs_FlexIndex 2) # []) ) "
```

I.e. we are able to compile tensor literals of this structure, this also works for multi-dimensional tensors...
In my understanding this happens because all the tensor literal calls end up in `ConstTensor` in `compileBuiltin` rather than in `NatTensorLiteral` etc.
Are these cases obsolete or are they triggered in another situation?

For reference here is the error message I get on the `dev` branch for the expression above (just in case that's relevant):
```bash
>cabal run vehicle -- compile itp --target Agda -s test2.vcl --output Test.agda
#Error in file 'test2.vcl' at Line 1, Columns 5-11: unable to work out a valid type for the overloaded expression 'TensorTC'.
#The possible options explored were:
#  1. Type
#    - rejected: unable to find a consistent type for the overloaded expression 'TensorTC'. In particular 'Rat' is not equal to '?23 Rat'.
#  2. Type
#    - rejected: unable to find a consistent type for the overloaded expression 'TensorTC'. In particular 'Rat' is not equal to 'Real'.
#  3. Type
#    - rejected: unable to find a consistent type for the overloaded expression 'TensorTC'. In particular 'Rat' is not equal to 'Bool'.
```

MatthewDaggitt replied:
Ah sorry, I meant (`Real` not `Rat`):
```
x : Tensor Real [2]
x = [1.0, 2.0]
```

MatthewDaggitt replied:
Will look at this in a couple of hours.

MatthewDaggitt replied:
You were absolutely right, `StackTensor` was not being evaluated where it should have been. I have fixed this in https://github.com/vehicle-lang/vehicle/pull/1038

samysweb replied:
Thank you, this allowed me to fix this: I am now generating 1-dimensional tensors and then stack them as needed... I saw you deactivated the ITP tests for `simple-vector` in #1038 -- I've followed suit and deactivated the (failing) Isabelle golden tests for this case. The other occurrences of tensors in the golden tests have been updated and still run with Isabelle compilation

samysweb replied:
so this should be resolved now (besides `simple-vector`)

MatthewDaggitt replied:
Yes, `simple-vector` needs fixing upstream. Our handling of zero-dimensional tensors is still slightly problematic.

MatthewDaggitt commented on vehicle/src/Vehicle/Backend/Isabelle/Compile.hs:
```diff
@@ -0,0 +1,897 @@
+          _ -> developerError $ "foreach/forall/exists tensor operations are currently only supports explicit lambda arguments with indexing type"
+        _ -> developerError $ "foreach/forall/exists tensor operations are currently only supports a single lambda argument"
+
+-- Default precedence for standard operations can be found at https://coq.inria.fr/doc/V8.18.0/refman/language/coq-library.html#notations
```
Incorrect reference?

samysweb replied:
removed

MatthewDaggitt commented on vehicle/src/Vehicle/Backend/Isabelle/Compile.hs:
```diff
@@ -0,0 +1,897 @@
+idxBasedOp :: (MonadIsabelleCompile m) => [LocaleDef] -> Code -> [Arg DecidabilityBuiltin] -> m Code
+idxBasedOp localeAssms op args = case args of
+        [(ExplicitArg _ _ (Lam _ binder _body))] -> case (typeOf binder) of
+          -- (App (Builtin _ IndexType) [maxIdx]) ->
```
Remove?

samysweb replied:
fixed

MatthewDaggitt reviewed (changes requested):
Apologies for the delay in reviewing. Looks great, a few minor comments, but nothing major!

samysweb committed:
Fix gitignore

samysweb committed:
Fix typo

samysweb committed:
Enabled tests and made them dependent on availability of Isabelle

samysweb committed:
removed stale comments

samysweb committed:
Empty lines between declarations

samysweb committed:
Addressing some of the code review comments:

- stylistic improvements
- unifying error message for universe types
- todo for records

samysweb committed:
Merge branch 'samysweb/vehicle-isabelle' of github.com:samysweb/vehicle into samysweb/vehicle-isabelle

MatthewDaggitt commented:
@samysweb I'm very sorry, there's going to be a small merge conflict with https://github.com/vehicle-lang/vehicle/pull/1022. In particular:

1.  I've removed the `DefRecord` constructor as it is now merged into `DefFunction`. As you can see in the changes to the Rocq and Agda files in that PR, the information about the original form of the definition is now stored in the `funSort` argument to `DefFunction` which you can then case match on.

2. I've added the ability to parameterise records by types. However, this purely for internal use only at the moment, so only unparameterised records should make it through to the ITP backends. Feel free to add support if you want (/can) but at the moment for the Agda/Rocq backends I'm just throwing an error if the parameter telescope is non-empty.

samysweb committed:
Fixed capitalization of true/false literals

samysweb committed:
Merge branch 'dev' into samysweb/vehicle-isabelle

samysweb committed:
Fixed version regarding records. This works for everything except record member access

samysweb commented on vehicle/src/Vehicle/Backend/Isabelle/Compile.hs:
```diff
@@ -0,0 +1,900 @@
+      return $ "let" <+> cBoundExpr <+> "in" <+> cBody
+    Lam _ binder body -> compileLam localeAssms binder body
+    Builtin _p b -> compileBuiltin isOutType localeAssms b []
+    App fun args -> compileApp isOutType localeAssms fun args
```
@MatthewDaggitt, it seems to me record member accesses/projections are now compiled to function applications? Is there any way for me to distinguish whether the function I am applying is a record member access/record projection? I need to be able to distinguish them for my compilation to work for the `simple-logic` example. Right now `myLogic.true` is seemingly treated like a function application of `true` to `myLogic`

P.S.: Does that mean the `RecordProj` case two lines further down is redundant/unused now?

MatthewDaggitt replied:
Ah yes right. This isn't a problem in the Agda or Rocq backends because this is actually how record projections are implemented. 

Given that it only affects the Isabelle backend I don't think we want to add this information to the `Identifier` class. Instead what I suggest is that you track which functions are record projections:

1. change `MonadIsabelleCompile` to include a `MonadState (Set Identifier) m` constraint. 
2. in `compileDecl` when you encounter a projection function, call `modify (Set.insert ident)` to add the identifier to the list of projection identifiers.
3. when you come across an `Identifier` when compiling an expression call `isProjectionFn <- gets (Set.member ident)`

Sorry that code may need a tiny bit of fiddling to get the argument order right, but it should morally work! Even better would be to abstract `isProjectionFn` into it's own method declared underneath `MonadIsabelleCompile` called `isProjectionFunction`.

MatthewDaggitt replied:
Yes, `RecordProj` should only be used inside the projection functions so you don't need to handle it. I should probably write a clean-up pass before the individual ITP compilation to avoid all these cases that we are guaranteed not to need...

samysweb committed:
Merge branch 'dev' into samysweb/vehicle-isabelle

samysweb committed:
Updated vehicle version in tests

samysweb committed:
Fixed compatibility of AnnInstance

samysweb committed:
Added tensor literal compilation (+ updated golden tests)

MatthewDaggitt commented:
I've also given you write access to the repository which is long overdue! Apologies.

samysweb committed:
Made record access work again, eliminated spurious empty lines

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

samysweb committed:
Adjusted readme

samysweb committed:
Fixed accidentally broken windController test invocations

samysweb committed:
Ormolu cleanup

samysweb committed:
Merge branch 'samysweb/vehicle-isabelle' of github.com:samysweb/vehicle into samysweb/vehicle-isabelle

samysweb committed:
Fixing additional ormolu issues

samysweb committed:
Github CI try 1

samysweb committed:
Briefly activate to test

samysweb committed:
Github CI try 2

samysweb committed:
Github CI try 3

samysweb committed:
Github CI try 4

samysweb committed:
Github CI try 5 (trying to get Caching to work)

samysweb committed:
Only run isabelle tests here

samysweb committed:
Fix caching condition

samysweb committed:
Trying mac another time

samysweb committed:
Tried to fix trailing white spaces manually

samysweb committed:
Fixing end of file new line

samysweb committed:
Merge branch 'dev' into samysweb/vehicle-isabelle

samysweb commented:
@MatthewDaggitt I implemented the record field access as you suggested and figured out a way to bring Isabelle into the CI (with some inspiration from the CI of another project I stumbled across [1]).
I've also merged with the up to date dev branch. From my side this is ready to merge (pending the succesful completion of the CI)

[1] https://github.com/kappelmann/Isabelle-Set
